File: lw7/slides/shapes/Triangle.cpp
#include "Triangle.h"
#include <algorithm>

CTriangle::CTriangle(const PointD& vertex1, const PointD& vertex2, const PointD& vertex3)
	: m_vertex1(vertex1)
	, m_vertex2(vertex2)
	, m_vertex3(vertex3)
{
}

std::optional<RectD> CTriangle::GetFrame()
{
	double minX = std::min({ m_vertex1.x, m_vertex2.x, m_vertex3.x });
	double minY = std::min({ m_vertex1.y, m_vertex2.y, m_vertex3.y });

	double maxX = std::max({ m_vertex1.x, m_vertex2.x, m_vertex3.x });
	double maxY = std::max({ m_vertex1.y, m_vertex2.y, m_vertex3.y });

	return RectD{ minX, minY, maxX - minX, maxY - minY };
}

void CTriangle::SetFrame(const RectD& rect)
{
	const auto currentFrame = *GetFrame();

	std::function updateVertexFn = [&](PointD& vertex) {
		const auto scaleX = (vertex.x - currentFrame.left) / currentFrame.width;
		const auto scaleY = (vertex.y - currentFrame.top) / currentFrame.height;

		vertex = PointD{ rect.left + rect.width * scaleX, rect.top + rect.height * scaleY };
	};

	updateVertexFn(m_vertex1);
	updateVertexFn(m_vertex2);
	updateVertexFn(m_vertex3);
}

void CTriangle::DrawBehaviour(ICanvas& canvas) const
{
	if (*GetOutlineStyle()->IsEnabled())
	{
		canvas.DrawLine(m_vertex1, m_vertex2);
		canvas.DrawLine(m_vertex2, m_vertex3);
		canvas.DrawLine(m_vertex3, m_vertex1);
	}

	if (*GetFillStyle()->IsEnabled())
	{
		canvas.FillPolygon(std::vector<PointD>{ m_vertex1, m_vertex2, m_vertex3 });
	}
}

File: lw9/src/model/shapes/Triangle.js
import {Shape} from "./Shape"
import {Point} from "../../common/Types"
import {Rect} from "../../common/Types"

class Triangle extends Shape {
    /**
     * @param {Rect} rect
     */
    constructor(rect) {
        super()

        /**
         * @type {Point}
         * @private
         */
        this._vertex1 = { x: rect.left, y: rect.top + rect.height }

        /**
         * @type {Point}
         * @private
         */
        this._vertex2 = { x: rect.left + rect.width / 2, y: rect.top }

        /**
         * @type {Point}
         * @private
         */
        this._vertex3 = { x: rect.left + rect.width, y: rect.top + rect.height }
    }

    /**
     * @override
     * @private
     */
    _getTypeImpl() {
        return 'triangle'
    }

    /**
     * @override
     * @private
     */
    _getFrameImpl() {
        const minX = Math.min(this._vertex1.x, this._vertex2.x, this._vertex3.x)
        const minY = Math.min(this._vertex1.y, this._vertex2.y, this._vertex3.y)

        const maxX = Math.max(this._vertex1.x, this._vertex2.x, this._vertex3.x)
        const maxY = Math.max(this._vertex1.y, this._vertex2.y, this._vertex3.y)

        return {
            left: minX,
            top: minY,
            width: maxX - minX,
            height: maxY - minY,
        }
    }

    /**
     * @override
     * @private
     */
    _setFrameImpl(rect) {
        const currentFrame = this._getFrameImpl()

        /**
         * @param {Point} vertex
         */
        function updateVertex(vertex) {
            const scaleX = (vertex.x - currentFrame.left) / currentFrame.width
            const scaleY = (vertex.y - currentFrame.top) / currentFrame.height

            vertex.x = rect.left + rect.width * scaleX
            vertex.y = rect.top + rect.height * scaleY
        }

        updateVertex(this._vertex1)
        updateVertex(this._vertex2)
        updateVertex(this._vertex3)
    }
}

export {
    Triangle,
}